lib/bap_future/bap_future.ml

```diff
@@ -0,0 +1,503 @@
+
+    let apply fs xs = merge fs xs ~f:(fun f x -> f x)
+
+    let upon finished s : 'a future =
+      let result = ref None in
+      observe s (fun x -> result := Some x);
+      let future,promise = Future.create () in
+      Future.upon finished (fun () -> match !result with
+          | None -> ()
+          | Some x -> Promise.fulfill promise x);
+      future
+
+    let before event xs = 
+      let future, promise = Future.create () in
+      let buf = Queue.create () in     
+      let id = subscribe xs (Queue.enqueue buf) in
+      Future.upon event
+        (fun () -> 
+           unsubscribe xs id;
+           Promise.fulfill promise (Queue.to_list buf)); 
+      future
+
+    let last_before e xs n =
+      let buf = Queue.create () in
+      let add_value x = 
+        if Queue.length buf >= n then 
+          ignore (Queue.dequeue buf);
+        Queue.enqueue buf x in
+      let id = subscribe xs add_value in
+      let future, promise = Future.create () in
+      Future.upon e 
+        (fun () -> 
+           unsubscribe xs id;
+           Promise.fulfill promise (Queue.to_list buf));
+      future
+
+    let take xs n = 
+      let buf = Queue.create () in 
+      let future, promise = Future.create () in
+      let add_value id x = 
+        if Queue.length buf < n then Queue.enqueue buf x;
+        if Queue.length buf = n then
+          (unsubscribe xs id;
+           Promise.fulfill promise (Queue.to_list buf)) in
+      watch xs add_value;
+      future
+
+    let nth xs n = 
+      let cnt = ref 0 in
+      let future, promise = Future.create () in
+      let f id x = 
+        if !cnt = n then 
+          (unsubscribe xs id;
+           Promise.fulfill promise x);
+        cnt := Int.succ !cnt; in
+      watch xs f;
+      future
+      
+    let find xs ~f = 
+      let future, promise = Future.create () in
+      let f id x = 
+        if f x then 
+          (unsubscribe xs id;
+           Promise.fulfill promise x) in
+      watch xs f;
+      future
+
+    let find_map xs ~f = 
+      let future, promise = Future.create () in
+      let f id x = match f x with
+        | None -> ()
+        | Some y ->
+          unsubscribe xs id;
+          Promise.fulfill promise y in
+      watch xs f;
+      future
+
+    let hd xs = 
+      let future, promise = Future.create () in
+      let f id x = 
+        unsubscribe xs id;
+        Promise.fulfill promise x in
+      watch xs f;
+      future
+
+    let tl xs =
+      let s, Signal publish = create () in
+      let ign = ref true in
+      let f x = 
+        if !ign then ign := false
+        else publish x in
+      link xs s f;
+      s
+
+    let foldw ?(stride = 1) xs width ~init ~f = 
+      let s, Signal publish = create () in
+      let nxt = ref None in
+      let res = ref init in
+      let cnt = ref Int.zero in
+      let try_push x = 
+        cnt := Int.succ !cnt;
+        res := f !res x;
+        if !cnt = width then
+          (cnt := Int.zero;
```

Comment:
Instead of using ugly parentheses, it is better to use `begin/end`, but this is still quite ugly, but good enough for a second grader. A usual functional approach is to abstract such code of blocks with lambda abstraction, e.g.,
```
... 
let reset () = 
    cnt := 0;
    nxt := 0; 
    publish !res in
...
if cnt.contents = width then reset ();
```

Even better approach would be to create a type `foldw_state` and have a record of state words, instead of spreading them across the function body.